### ISSUE TITLE
Kestrel Picks the Wrong SSL Certificate

### ISSUE THREAD

**Issue by keggster101020** [dotnet/aspnetcore#51688]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I have multiple certificates installed on a remote machine that have similar subject names. Kestrel seems to be using a substring matching when selecting the certificate resulting in it picking the wrong certificate to bind for SSL. 

Example:

- Cert1: CN = "a.b.c.d.e.com" 
  - issued after Cert2
- Cert2: CN = "c.d.e.com"

In appsettings.json:

```json
"Kestrel": {
	"Certificates": {
		"Default": {
			"Subject": "c.d.e.com",
			"Store": "My",
			"Location": "LocalMachine"
		}
	}
}
```

Whenever I navigate to the server in a browser, I get an SSL trust error. Looking at the certificate in the browser, I see the cert1 has been bound to 443 (CN: a.b.c.d.e.com) instead of cert2. 

### Expected Behavior

The certificate whose subject is exactly matching what I put in appsettings.json is used.

### .NET Version

net6.0

### Anything else?

This is being deployed in a service fabric cluster hosted in Azure. 

Key vault extension for windows is being used to pull the certificates from key vault onto the machine with the `link on renewal` property set to false.

**Comment by Tratcher** [comment]
Here's the code for that:
https://github.com/dotnet/aspnetcore/blob/6c1b3dfb66a1b66cc32ee26bbc23f6472f1dc985/src/Servers/Kestrel/Core/src/CertificateLoader.cs#L23-L54
https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509findtype?view=net-7.0

> FindBySubjectName | 1 | The findValue parameter for the Find(X509FindType, Object, Boolean) method must be a string representing the subject name of the certificate. This is a less specific search than that provided by the FindBySubjectDistinguishedName enumeration value. Using the FindBySubjectName value, the Find(X509FindType, Object, Boolean) method performs a case-insensitive string comparison using the supplied value. For example, if you pass "MyCert" to the Find(X509FindType, Object, Boolean) method, it will find all certificates with the subject name containing that string, regardless of other subject values. Searching by distinguished name is a more precise search.

Find seems to return all matches, but then we check for a more specific match. Is SimpleName the wrong thing to match?

**Comment by amcasey** [comment]
This looks like https://github.com/dotnet/aspnetcore/issues/49062

**Comment by keggster101020** [comment]
@amcasey this does look like my issue as well since I'm able to work around my problem by issuing a new version of the cert I want to use as my SSL certificate. 

Obviously, this is a temporary fix; once the other cert gets a new version SSL trust will break once again. 

It looks like this change is slated for .NET 9 so I will try the workaround in #49062 and pass the certificate to Kestrel when building the web host.

**Comment by amcasey** [comment]
Yep, unfortunately, the fix came in really late in the 8.0 cycle and it didn't have enough upvotes to justify shipping it after feature complete.  If the workaround doesn't solve your problem, please let us know by filing a new issue referencing #49062 and the fact that you're on 6.0.  Thanks!
